tsdb/head.go

```diff
@@ -571,31 +620,106 @@ func (h *Head) Init(minValidTime int64) error {
 		return nil
 	}
 
-	level.Info(h.logger).Log("msg", "replaying WAL, this may take awhile")
-	// Backfill the checkpoint first if it exists.
-	dir, startFrom, err := wal.LastCheckpoint(h.wal.Dir())
-	if err != nil && err != record.ErrNotFound {
-		return errors.Wrap(err, "find last checkpoint")
-	}
+	refSeries := map[uint64]*memSeries{}
+	level.Info(h.logger).Log("msg", "creating series from WAL")
 	multiRef := map[uint64]uint64{}
-	if err == nil {
-		sr, err := wal.NewSegmentsReader(dir)
-		if err != nil {
-			return errors.Wrap(err, "open checkpoint")
+	// Backfill series from the checkpoint first if it exists.
+	if err := h.executeOnCheckpointReader(func(checkpointReader *wal.Reader) error {
+		// A corrupted checkpoint is a hard error for now and requires user
```

Comment:
We should treat it like any other wal corruption, delete what is corrupted onwards and continue to startup

Reply:
This is the same old comment what was present, haven't changed anything here in terms of processing the checkpoint. Did you mean for the chunk files that we should delete anything after the corruption?

Reply:
Yes

Reply:
So the corruption in the chunks files is not obvious. There are no records to decode specifically that we can error on. The error is more like `"not enough bytes to read the next chunk"`. In that case, we are not even sure if all the chunks read till that part is correct, so we have to discard all the chunks from that file, which sounds tricky after having parsed all the (right/wrong) chunks till then. So a safe fallback that I see is to discard all the chunks that we read till now and recover entirely from WAL.